src/bin/pg_dump/dump_babel_utils.c

```diff
@@ -993,47 +974,70 @@ prepareForBabelfishDatabaseDump(Archive *fout, SimpleStringList *schema_include_
 void
 setBabelfishDependenciesForLogicalDatabaseDump(Archive *fout)
 {
+	int        		i;
 	PQExpBuffer		query;
 	PGresult		*res;
-	TableInfo		*extprop_table;
-	TableInfo		*sysdb_table;
-	TableInfo		*namespace_ext_table;
-	TableInfo		*schema_perms_table;
-	DumpableObject		*dobj;
-	DumpableObject		*refdobj;
+	TableInfo		*sysdb_table = NULL;
+	DumpableObject		*refdobj = NULL;
 
 	if (!isBabelfishDatabase(fout) || fout->dopt->binary_upgrade || fout->dopt->dataOnly)
 		return;
 
 	query = createPQExpBuffer();
-	/* get oids of sys.babelfish_sysdatabases, sys.babelfish_namespace_ext, babelfish_extended_properties, babelfish_schema_permissions tables */
+	/*
+	 * Get OIDs for following desired catalog tables:
+	 * sys.babelfish_sysdatabases
+	 * sys.babelfish_namespace_ext
+	 * babelfish_extended_properties
+	 * babelfish_schema_permissions
+	 */
 	appendPQExpBufferStr(query,
 						 "SELECT oid "
 						 "FROM pg_class "
 						 "WHERE relname in ('babelfish_sysdatabases', "
-						 "'babelfish_schema_permissions', 'babelfish_namespace_ext', 'babelfish_extended_properties') "
+						 "'babelfish_schema_permissions', 'babelfish_namespace_ext', "
+						 "'babelfish_extended_properties') "
 						 "AND relnamespace = 'sys'::regnamespace "
 						 "ORDER BY relname;");
 	res = ExecuteSqlQuery(fout, query->data, PGRES_TUPLES_OK);
 
-	Assert(PQntuples(res) == 4);
-	extprop_table = findTableByOid(atooid(PQgetvalue(res, 0, 0)));
-	namespace_ext_table = findTableByOid(atooid(PQgetvalue(res, 1, 0)));
-	schema_perms_table = findTableByOid(atooid(PQgetvalue(res, 2, 0)));
-	sysdb_table = findTableByOid(atooid(PQgetvalue(res, 3, 0)));
-	Assert(sysdb_table != NULL && namespace_ext_table != NULL && extprop_table != NULL && schema_perms_table != NULL);
-	refdobj = (DumpableObject *) sysdb_table->dataObj;
 	/*
-	 * Make babelfish_schema_permissions, babelfish_namespace_ext and babelfish_extended_properties tables dependent upon
-	 * babelfish_sysdatabases table so that we dump babelfish_sysdatabases table's data before both of them.
-	 * This is needed to generate and handle new "dbid" during logical database restore.
+	 * babelfish_sysdatabases, babelfish_namespace_ext and babelfish_extended_properties
+	 * catalogs are old enough that there must be atleast 3 rows returned.
+	 */
+	Assert(PQntuples(res) >= 3);
+
+	/* First iterate through all the rows to find out the OID of babelfish_sysdatabases table */
+	for (i = 0; i < PQntuples(res); i++)
+	{
+		TableInfo *tab = findTableByOid(atooid(PQgetvalue(res, i, 0)));
+
+		if (pg_strcasecmp(tab->dobj.name, "babelfish_sysdatabases") == 0)
```

Comment:
I would suggest to execute two SQL query, first one to find `babelfish_sysdatabases` and second one to find the all other catalogs to avoid string comparison we are doing to search for it. Or maybe include a flag in same query to identify it.

Reply:
My initial idea was to reduce the number of database calls and get the OIDs with the same query but since this function only gets executed once during the whole dump process, having two separate queries also don't hurt much. So suggestion taken.